DEPR: group by one element list gets scalar keys

natmokval opened the pull request:
- [x] closes #58858

changed the behavior: group `keys` to be `1-tuples` instead of `scalar` when grouping with a `length-1 list`

natmokval committed:
deprecate group by one element list get scalar keys

natmokval committed:
fix tests

natmokval committed:
correct docstring

natmokval committed:
correct docs

natmokval commented:
I deprecated the behavior for `groups`: group `keys `are now 1-tuples instead of scalars when grouping with a length-1 list.
I fixed the tests except for this one `test_groupby_raises_category_on_category` in `pandas/tests/groupby/test_raises.py`. @rhshadrach, could you please advise me on how to fix this test?

natmokval committed:
revert the changes of the group's behavior

rhshadrach commented on pandas/core/groupby/grouper.py:
```diff
@@ -667,6 +671,15 @@ def groups(self) -> dict[Hashable, Index]:
         codes, uniques = self._codes_and_uniques
         uniques = Index._with_infer(uniques, name=self.name)
         cats = Categorical.from_codes(codes, uniques, validate=False)
+        if not self.key_dtype_str:
+            warnings.warn(
+                "`groups` by one element list returns scalar is deprecated "
+                "and will be removed. In a future version `groups` by one element "
+                "list will return tuple. Use ``df.groupby(by='a').groups`` "
+                "instead of ``df.groupby(by=['a']).groups`` to avoid this warning",
+                FutureWarning,
+                stacklevel=find_stack_level(),
+            )
```
Can you move this to `GroupBy.groups`? Then you won't need to modify `Grouping` (e.g. include extra arguments) at all.

rhshadrach commented:
> @rhshadrach, could you please advise me on how to fix this test?

Sorry for the delay! I believe you've resolved this, let me know if not.

natmokval committed:
move the FutureWarning to GroupBy.groups

natmokval committed:
Merge branch 'main' into depr-group-by-1dim-list-get-scalar-keys

natmokval commented:
> Sorry for the delay! I believe you've resolved this, let me know if not.

Thanks @rhshadrach for reviewing this PR. I moved the FutureWarning from Grouping to GroupBy.groups as you suggested and added a note to v3.0.0. Could you please take a look at my changes?

rhshadrach commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2996,3 +2996,15 @@ def test_groupby_multi_index_codes():
     tm.assert_index_equal(index, MultiIndex.from_frame(index.to_frame()))
+
+
+def test_groupby_keys_1length_list():
```
Since this behavior is already hit in tests, I don't think we need to add a new test here.

natmokval replied:
thanks, I removed the new test

natmokval committed:
remove unnecessary test, rename a variable

natmokval committed:
Merge branch 'main' into depr-group-by-1dim-list-get-scalar-keys

natmokval commented on pandas/tests/groupby/test_grouping.py:
```diff
@@ -719,15 +732,18 @@ def test_grouping_labels(self, multiindex_dataframe_random_data):
         expected = {Timestamp("2011-01-01"): Index(list(range(364)))}
-        tm.assert_dict_equal(result.groups, expected)
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            result = result.groups
```
thanks, done

rhshadrach reviewed (approved):
lgtm

rhshadrach commented:
Thanks @natmokval!
